File: npm/src/adapters/codex.ts
// Codex (OpenAI Codex CLI) adapter for the npm-launched flow.
//
// Codex reads MCP server registrations from `~/.codex/config.toml`
// under `[mcp_servers.<name>]` tables. The file is hand-edited by
// users (it also holds non-MCP settings like model defaults), so we
// must not round-trip it through a generic TOML library — that would
// drop comments, blank lines, and stable key ordering.
//
// We instead rely on `./toml.ts`, a hand-rolled minimal reader/writer
// scoped to the Codex MCP subset. It locates the [mcp_servers.<name>]
// block (and any [mcp_servers.<name>.env] sub-table) by line scan
// and rewrites only that slice, preserving every other byte of the
// file.
//
// Detection: we test for the existence of the `~/.codex` directory
// rather than the file itself. Codex creates `config.toml` lazily
// on first config write, so requiring the file would incorrectly
// report Codex as "not installed" for fresh installations.

import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";

import type { Adapter, AdapterConfig } from "./types.js";
import { backupFile, restoreFromBackup } from "./fileutil.js";
import { removeMcpServer, upsertMcpServer } from "./toml.js";

export class CodexAdapter implements Adapter {
  name(): string {
    return "codex";
  }

  private codexDir(): string {
    return join(homedir(), ".codex");
  }

  private configPath(): string {
    return join(this.codexDir(), "config.toml");
  }

  installed(): boolean {
    // The directory is the install marker — config.toml may not
    // exist yet on a fresh Codex setup.
    return existsSync(this.codexDir());
  }

  async install(cfg: AdapterConfig): Promise<void> {
    const path = this.configPath();
    const prior = existsSync(path) ? readFileSync(path, "utf8") : "";

    // Only back up when there's something to restore. backupFile is
    // already a no-op for missing paths, but checking explicitly
    // keeps the intent clear and avoids polluting the directory
    // with empty backup files for new installs.
    if (existsSync(path)) backupFile(path);

    let next = upsertMcpServer(prior, "klio", {
      command: "docker",
      args: ["exec", "-i", cfg.bridgeContainer, "klio-mcp"],
      env: cfg.env,
    });

    // Pre-approve klio's MCP tools so Codex doesn't prompt the user
    // before each invocation. Codex's per-server approval-mode lives
    // under [apps.<server>] (sibling of [mcp_servers.<server>]):
    //
    //   [apps.klio]
    //   default_tools_approval_mode = "auto"
    //
    // Without it, the user sees an approval modal the first time
    // each of klio's 7 tools (recall/remember/observe/plan/decide/
    // note/space) fires — defeats the "memory just works" promise.
    //
    // Idempotent: appended only when not already present. A no-op
    // when Codex's TOML parser doesn't recognise the key (older
    // builds), since unknown TOML keys are silently ignored.
    next = ensureAppsKlioAutoApproval(next);

    writeFileSync(path, next, { mode: 0o644 });
  }

  async uninstall(): Promise<void> {
    const path = this.configPath();
    if (!existsSync(path)) return;

    try {
      restoreFromBackup(path);
      return;
    } catch {
      // No backup found — fall through to in-place strip. Partial
      // cleanup beats a hard failure when the user has pruned
      // backup files.
    }

    let stripped = removeMcpServer(readFileSync(path, "utf8"), "klio");
    stripped = removeAppsKlioAutoApproval(stripped);
    writeFileSync(path, stripped, { mode: 0o644 });
  }
}

/**
 * Append `[apps.klio]` with `default_tools_approval_mode = "auto"` to
 * a Codex config body if it isn't already present. Idempotent — re-
 * running install on the same body is a byte-identical no-op.
 *
 * We check via literal substring rather than parsing because:
 *   - the toml.ts editor only knows the [mcp_servers.<name>] subtree,
 *   - a user might have edited the block by hand (different
 *     formatting, additional sub-keys),
 *   - any presence of `[apps.klio]` means our intent is already
 *     declared even if values differ; respect that.
 */
function ensureAppsKlioAutoApproval(body: string): string {
  if (/\[apps\.klio\]/.test(body)) return body;
  // Bare block — no leading comment header. The TOML editor's
  // `upsertMcpServer` replaces the slice from `[mcp_servers.klio]`
  // up to the NEXT `[...]` header, so any comment lines between the
  // mcp_servers block and `[apps.klio]` would be stomped on the
  // second install (breaking idempotency). Keeping the block bare
  // means `findBlockRange`'s slice ends exactly at `[apps.klio]`,
  // making repeat installs byte-identical.
  const trailer = body.length === 0 || body.endsWith("\n") ? "" : "\n";
  return (
    body +
    trailer +
    "[apps.klio]\n" +
    'default_tools_approval_mode = "auto"\n'
  );
}

/**
 * Strip the [apps.klio] block (and any sub-keys until the next
 * top-level [...] header or EOF). Mirrors the conservative line-
 * scan style of toml.ts to keep peer content byte-stable.
 */
function removeAppsKlioAutoApproval(body: string): string {
  const lines = body.split("\n");
  const startIdx = lines.findIndex((l) => l.trim() === "[apps.klio]");
  if (startIdx < 0) return body;

  let endIdx = lines.length;
  for (let i = startIdx + 1; i < lines.length; i++) {
    const t = lines[i].trim();
    if (t.startsWith("[") && !t.startsWith("[apps.klio.")) {
      endIdx = i;
      break;
    }
  }

  // Also drop our own header comment block that immediately precedes
  // the [apps.klio] line (we wrote 3 commented-out lines above the
  // table header in `ensureAppsKlioAutoApproval`). Walk backwards
  // while lines start with `#` AND mention "klio".
  let trueStart = startIdx;
  while (
    trueStart > 0 &&
    lines[trueStart - 1].trimStart().startsWith("#") &&
    /klio/i.test(lines[trueStart - 1])
  ) {
    trueStart--;
  }
  // Also drop one blank line before the header block, if present, so
  // we don't leak an extra newline gap on uninstall.
  if (trueStart > 0 && lines[trueStart - 1].trim() === "") {
    trueStart--;
  }

  const before = lines.slice(0, trueStart);
  const after = lines.slice(endIdx);
  return [...before, ...after].join("\n");
}

File: bridge/internal/agentadapters/codex.go
package agentadapters

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	toml "github.com/pelletier/go-toml/v2"
)

// CodexAdapter wires Klio into OpenAI Codex by patching
// ~/.codex/config.toml with the Klio MCP server entry. Codex — like
// Cursor and unlike Claude Code — does not expose a hooks system, so
// this adapter only manages MCP config; session capture for Codex
// users still flows through the Klio MCP tools (recall/remember/...)
// which Codex invokes directly.
//
// Config file shape (Codex's published TOML format):
//
//	[mcp_servers.klio]
//	command = "/abs/path/to/klio-mcp"
//	args = []
//
//	[mcp_servers.klio.env]
//	KLIO_SOCKET_PATH = "..."
//
//	[mcp_servers.<other-server>]
//	command = "..."
//
// We patch in place so any non-Klio servers and unrelated top-level
// keys (e.g. `telemetry = false`) the user has set stay intact, and we
// back up the file before writing for the same reason as Cursor: the
// user may have hand-edited it.
type CodexAdapter struct{}

func NewCodexAdapter() *CodexAdapter { return &CodexAdapter{} }

func (a *CodexAdapter) Name() string { return "codex" }

// configPath returns the user-scoped Codex config file location. Codex
// reads MCP servers from this single location; klio init runs at the
// user level and registers globally so Klio works in every project the
// user opens.
func (a *CodexAdapter) configPath() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".codex", "config.toml")
}

// Installed reports true when Codex's user-scoped config directory
// (~/.codex) exists. We deliberately do NOT require config.toml itself
// to exist — fresh Codex installs may not create it until first use.
// Detecting the directory is enough to say "Codex is on this machine"
// and start managing its MCP config.
func (a *CodexAdapter) Installed() bool {
	home, err := os.UserHomeDir()
	if err != nil {
		return false
	}
	if _, err := os.Stat(filepath.Join(home, ".codex")); err == nil {
		return true
	}
	return false
}

// Install patches ~/.codex/config.toml to register the Klio MCP server.
// Idempotent: re-running with identical Config replaces the prior
// `[mcp_servers.klio]` section without disturbing peer entries or
// unrelated top-level keys.
func (a *CodexAdapter) Install(cfg Config) error {
	if cfg.KlioMcpBinary == "" {
		return errors.New("agentadapters: Config.KlioMcpBinary is required")
	}

	path := a.configPath()
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return fmt.Errorf("create %s: %w", filepath.Dir(path), err)
	}

	// If the file does not exist, treat it as an empty document. Unlike
	// Cursor's mcp.json (where we seed `{}` so the read path is
	// uniform), TOML's empty-document representation is just an empty
	// file, which readTOML already handles. We only seed the file when
	// it's missing so that backupFile (which requires the file to
	// exist) has something to copy.
	existed := true
	if _, err := os.Stat(path); errors.Is(err, os.ErrNotExist) {
		existed = false
		if err := os.WriteFile(path, []byte(""), 0o644); err != nil {
			return fmt.Errorf("seed %s: %w", path, err)
		}
	}

	settings, err := readTOML(path)
	if err != nil {
		return err
	}
	if existed {
		if err := backupFile(path); err != nil {
			return fmt.Errorf("backup %s: %w", path, err)
		}
	}

	servers, _ := settings["mcp_servers"].(map[string]any)
	if servers == nil {
		servers = map[string]any{}
	}

	entry := map[string]any{
		"command": cfg.KlioMcpBinary,
		"args":    []any{},
	}
	if env := envToMap(cfg.Env); env != nil {
		entry["env"] = env
	}
	servers["klio"] = entry
	settings["mcp_servers"] = servers

	return writeTOML(path, settings)
}

// Uninstall restores ~/.codex/config.toml from the most recent backup.
// If no backup exists (e.g. user manually deleted it, or the original
// install pre-dated the backup convention), we fall back to removing
// the `[mcp_servers.klio]` section from the current file rather than
// failing — partial cleanup is better than no cleanup.
func (a *CodexAdapter) Uninstall() error {
	path := a.configPath()
	if _, err := os.Stat(path); errors.Is(err, os.ErrNotExist) {
		return nil
	}
	if err := restoreFromBackup(path); err == nil {
		return nil
	}
	// Backup missing — strip the klio entry in place.
	settings, err := readTOML(path)
	if err != nil {
		return err
	}
	if servers, ok := settings["mcp_servers"].(map[string]any); ok {
		delete(servers, "klio")
		if len(servers) == 0 {
			delete(settings, "mcp_servers")
		} else {
			settings["mcp_servers"] = servers
		}
	}
	return writeTOML(path, settings)
}

// readTOML loads a TOML file as a map[string]any. Empty/missing files
// return an empty map. Malformed TOML returns an error.
//
// Kept private to this file (rather than promoted to util.go) because
// Codex is currently the only TOML-backed adapter; if a second TOML
// adapter lands later, this is a one-line move.
func readTOML(path string) (map[string]any, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	if len(strings.TrimSpace(string(data))) == 0 {
		return map[string]any{}, nil
	}
	var out map[string]any
	if err := toml.Unmarshal(data, &out); err != nil {
		return nil, fmt.Errorf("config at %s is not valid TOML: %w", path, err)
	}
	return out, nil
}

// writeTOML marshals data back to the path. go-toml/v2 emits stable
// output (sorted keys at each level) which makes the idempotency
// property — re-installing with the same Config produces a
// byte-identical file — hold even after a round-trip through the
// in-memory map.
func writeTOML(path string, data map[string]any) error {
	body, err := toml.Marshal(data)
	if err != nil {
		return err
	}
	return os.WriteFile(path, body, 0o644)
}
